File: Striver/Trees/Q11.java
package Striver.Trees;

public class Q11{
    // Check if a binary tree is balanced or not

    public boolean isBalanced(Node root) {
        if(root == null) return true;
        return height(root) != -1;
    }

    public int height(Node root){
        if(root == null) return 0;
        int left = height(root.left);
        if(left == -1) return -1;
        int right = height(root.right);
        if(right == -1) return -1;
        if(Math.abs(left - right) > 1) return -1;
        return Math.max(left, right) + 1;
    }

    public static void main(String[] args) {
        Node root = new Node(10);
        root.left = new Node(20);
        root.left.left = new Node(40);
        root.left.right = new Node(50);
        root.right = new Node(30);
        root.right.right = new Node(70);
        root.right.right.right = new Node(80);
        Q11 q = new Q11();
        
        System.out.println(q.isBalanced(root));
    }
}

File: Striver/Trees/Q11.py
# check binary tree is balanced or not

from Node import Node

def height(root):
    if not root:
        return 0
    lh = height(root.left)
    if lh == -1:
        return -1
    rh = height(root.right)
    if rh == -1:
        return -1
    if abs(lh - rh) > 1:
        return -1
    return 1 + max(lh, rh)

def isBalanced(root):
    if not root:
        return True
    return height(root) != -1

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.left.left = Node(8)
root.left.left.right = Node(9)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

print(isBalanced(root))